src/repositories/updater.rs

```diff
@@ -0,0 +1,276 @@
+        metadata: &MetadataPackage,
+        config: Arc<Config>,
+        db: Pool,
+    ) -> Result<Option<i32>> {
+        let updater = match T::new(config, db)? {
+            Some(updater) => updater,
+            None => {
+                return Ok(None);
+            }
+        };
+        let repo = match &metadata.repository {
+            Some(url) => url,
+            None => {
+                debug!(
+                    "did not collect {} stats as no repository URL was present",
+                    T::name()
+                );
+                return Ok(None);
+            }
+        };
+        match updater.load_repository(conn, repo) {
+            Ok(repo) => Ok(repo),
+            Err(err) => {
+                warn!("failed to collect {} stats: {}", T::name(), err);
+                Ok(None)
+            }
+        }
+    }
+
+    pub fn start_crons(config: Arc<Config>, context: &dyn Context) -> Result<()> {
+        macro_rules! start_cron {
+            ($typ:ty) => {
+                if let Some(updater) = <$typ>::new(config.clone(), context.pool()?)? {
+                    cron(
+                        concat!(stringify!($typ), " stats updater"),
+                        Duration::from_secs(60 * 60),
+                        move || {
+                            updater.update_all_crates()?;
+                            Ok(())
+                        },
+                    )?;
+                } else {
+                    log::warn!(
+                        "{} stats updater not started as no token was provided",
+                        <$typ>::name()
+                    );
+                }
+            };
+        }
+
+        start_cron!(GithubUpdater);
+        start_cron!(GitlabUpdater);
+        Ok(())
+    }
+
+    pub fn update_all_crates(ctx: &dyn Context) -> Result<()> {
+        fn inner<T: Updater>(ctx: &dyn Context) -> Result<()> {
+            match T::new(ctx.config()?, ctx.pool()?)? {
+                Some(up) => up.update_all_crates()?,
+                None => warn!("missing {} token", T::name()),
+            }
+            Ok(())
+        }
+
+        inner::<GithubUpdater>(ctx)?;
+        inner::<GitlabUpdater>(ctx)?;
+        Ok(())
+    }
+
+    pub fn backfill_repositories(ctx: &dyn Context) -> Result<()> {
+        fn inner<T: Updater>(ctx: &dyn Context) -> Result<()> {
+            match T::new(ctx.config()?, ctx.pool()?)? {
+                Some(up) => up.backfill_repositories()?,
+                None => warn!("missing {} token", T::name()),
+            }
+            Ok(())
+        }
+
+        inner::<GithubUpdater>(ctx)?;
+        inner::<GitlabUpdater>(ctx)?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct RepositoryName<'a> {
+    pub owner: &'a str,
+    pub repo: &'a str,
+    pub host: &'a str,
+}
+
+pub fn get_icon_name(host: &str) -> &'static str {
+    if GithubUpdater::hosts().iter().any(|&h| h == host) {
+        "github"
+    } else {
```

Comment:
This should be checking gitlab's hosts too and warning if all checks fail for better handling of errors when modifying the updaters in the future.

Reply:
This function is only called in case we have information on the repository. And since we only provide such information for github and gitlab, this function is only called then.

Reply:
Yes, but if we change the set of hosts handled by `GitlabUpdater`, or add a new updater instance, without remembering to update this, it will use the "default" of `gitlab`, which could easily slip through testing.

Reply:
I see. Then I'll return an empty string in case it's not gitlab either.